src/routes/route-factory.js

```diff
@@ -2,41 +2,34 @@
 
 const express = require('express');
 const errors = require('@feathersjs/errors');
+
 const parseQueryParams = require('../middlewares/parse-query');
 const { hashPasswordOptional } = require('../middlewares/password-hasher');
+const createIdParamValidator = require('../middlewares/id-param-validator');
+const createPermissionVerifier = require('../middlewares/permission-verifier');
 
 module.exports = function createRoute(repository, permissions) {
   const router = express.Router();
 
+  const validateId = createIdParamValidator(repository);
+
+  const permissionVerifier = createPermissionVerifier(permissions);
+
   router.route('/')
-    .get(parseQueryParams, list)
-    .post(hashPasswordOptional, add);
+    .get(permissionVerifier.read, parseQueryParams, list)
+    .post(permissionVerifier.create, hashPasswordOptional, add);
 
   router.param('id', validateId);
 
   router.route('/:id')
-    .get(view)
-    .put(hashPasswordOptional, update)
-    .delete(remove);
-
-  function validateId(request, response, next, id) {
-    if (!id) return next(new errors.BadRequest('ID_NOT_RECEIVED'));
-
-    return repository.exists(id)
-      .then((result) => {
-        if (!result) return next(new errors.NotFound('ID_NOT_FOUND'));
-
-        return next();
-      })
-      .catch(next);
-  }
+    .get(permissionVerifier.read, view)
+    .put(permissionVerifier.update, hashPasswordOptional, update)
+    .delete(permissionVerifier.delete, remove);
 
   async function list(request, response, next) {
-    if (!permissions[request.token.userRole].read) return next(new errors.Forbidden('READ_PERMISSION_MISSING'));
-
-    return repository.list(request.query)
+    repository.list(request.query)
       .then((result) => {
-        if (result.count === 0) return next(new errors.NotFound('NO_RESULT'));
+        if (result.count === 0) return next(new errors.NotFound('NOT_FOUND'));

```

Comment:
for list you can have 0 result, you don't have to throw not found error

Reply:
should I still return the 404 status code?